protocol/messenger_handler.go

```diff
@@ -927,109 +927,113 @@ func (m *Messenger) HandleHistoryArchiveMagnetlinkMessage(state *ReceivedMessage
 			m.downloadHistoryArchiveTasksWaitGroup.Add(1)
 			go func() {
 				defer m.downloadHistoryArchiveTasksWaitGroup.Done()
-				downloadTaskInfo, err := m.communitiesManager.DownloadHistoryArchivesByMagnetlink(id, magnetlink)
-				if err != nil {
-					logMsg := "failed to download history archive data"
-					if err == communities.ErrTorrentTimedout {
-						m.communitiesManager.LogStdout("torrent has timed out, trying once more...")
-						downloadTaskInfo, err = m.communitiesManager.DownloadHistoryArchivesByMagnetlink(id, magnetlink)
-						if err != nil {
-							m.communitiesManager.LogStdout(logMsg, zap.Error(err))
-							return
-						}
+				m.downloadAndImportHistoryArchives(id, magnetlink)
+			}()
+
+			return m.communitiesManager.UpdateMagnetlinkMessageClock(id, clock)
+		}
+	}
+	return nil
+}
+
+func (m *Messenger) downloadAndImportHistoryArchives(id types.HexBytes, magnetlink string) {
+	downloadTaskInfo, err := m.communitiesManager.DownloadHistoryArchivesByMagnetlink(id, magnetlink)
+	if err != nil {
+		logMsg := "failed to download history archive data"
+		if err == communities.ErrTorrentTimedout {
+			m.communitiesManager.LogStdout("torrent has timed out, trying once more...")
+			downloadTaskInfo, err = m.communitiesManager.DownloadHistoryArchivesByMagnetlink(id, magnetlink)
+			if err != nil {
+				m.communitiesManager.LogStdout(logMsg, zap.Error(err))
+				return
+			}
+		} else {
+			m.communitiesManager.LogStdout(logMsg, zap.Error(err))
+			return
+		}
+	}
+
+	if downloadTaskInfo.Cancelled {
+		m.communitiesManager.LogStdout(fmt.Sprintf("downloaded %d of %d archives so far", downloadTaskInfo.TotalDownloadedArchivesCount, downloadTaskInfo.TotalArchivesCount))
+		return
+	}
+
+	cancel := m.communitiesManager.GetHistoryArchiveDownloadTask(id.String())
+	importTicker := time.NewTicker(100 * time.Millisecond)
+	defer importTicker.Stop()
+
+importMessageArchivesLoop:
+	for {
+		select {
+		case <-cancel:
+			m.communitiesManager.LogStdout("interrupted importing history archive messages")
+			return
+		case <-importTicker.C:
+
+			archiveIDsToImport, err := m.communitiesManager.GetMessageArchiveIDsToImport(id)
```

Comment:
Right, so generally we keep track of any history archive tasks per community such that, in needed, we can cancel them.

The way we do this is by keeping a list of channels that can emit a `cancel` signal. For example, you can see it being retrieved here, a few lines above: https://github.com/status-im/status-go/pull/3013/files#diff-59d196f7f72de56c64ba2c559fcbb51762c418a22d2b83c59662b772a09b3c0aR961

Here's roughly how it works:

1. When we receive a magnetlink, we and there's already a `historyArchiveDownloadTask` for this community, we'll take it and emit a cancel signal on it.  Then, we create a new one for the currently processed magnet link messages
2. The `cancel` channel is used to:
2.1. Cancelling the download of the community archive index
2.2. Cancelling the download of community archive data
2.3 Cancelling the import of download archive data into the database

So at any stage, when there's multiple magnetlink messages for the same community coming in, every message will first mark any previous download/import task for that community as cancelled before proceeding.

So there should only ever be one download/import task in progress.

Reply:
Suppose we have two goroutines A and B, A executes to line 973 (message_handler.go), then B executes to line 2138 (manager.go) `close(stop)`, B won't block after executing close and won't wait for A to finish, then it is possible that B and A both enter the same area(line 973)? @0x-r4bbit

Reply:
Ah I see.

So yea, I think it's theory it's possible. In practice I'd say it's unlikely though. Assuming the scenario you've described above, even though B won't block, it would have to be very fast (downloading archive data) to reach this line before A hits `cancel`

But I see where you're coming from. To be 100% sure, we'd have to block `B` until we know for sure that `A` has finished cancelling.

Reply:
I've made the necessary changes in #3012 
This PR has been updated accordingly 

TLDR:

It checks if there's an ongoing download task via and cancels it using its attached cancel channel. It then piggy bags on the same mechanism where we wait for cancellation to finish until we continue

Reply:
nice job🫶, i'm going to approve